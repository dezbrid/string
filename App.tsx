import React, {useEffect, useState} from 'react';
import {SafeAreaView, StatusBar} from 'react-native';
import BarText from '@components/BarText';
import MessageView from '@components/MessageView';

const PARENTHESES = 'p';
const BRACKETS = 'b';
const CURLY_BRACKETS = 'c';

const CHART_OPEN_PARENTHESES = '(';
const CHART_OPEN_BRACKETS = '[';
const CHART_OPEN_CURLY_BRACKETS = '{';
const CHART_CLOSE_PARENTHESES = ')';
const CHART_CLOSE_BRACKETS = ']';
const CHART_CLOSE_CURLY_BRACKETS = '}';

interface Chart {
  type: string;
  originalPosition: number;
}
const App = () => {
  const [text, setText] = useState<string>('');
  const [message, setMessage] = useState<string>('');
  const [openChart, setOpenChart] = useState<Chart[]>([]);
  const [closeChart, setCloseChart] = useState<Chart[]>([]);

  useEffect(() => {
    setMessage('');

    let arrayOpen: Chart[] = [];
    let arrayClose: Chart[] = [];
    if (openChart.length > closeChart.length) {
      setMessage('Error, falta caracter de cierre');
    }
    if (openChart.length < closeChart.length) {
      setMessage('Error, falta caracter de apertura');
    }
    if (openChart.length === closeChart.length) {
      for (const index in openChart) {
        if (openChart[index].type !== closeChart[index].type) {
          arrayOpen = [...arrayOpen, openChart[index]];
          arrayClose = [...arrayClose, closeChart[index]];
        } else if (
          closeChart[index].originalPosition < openChart[index].originalPosition
        ) {
          setMessage('Error, tienes algun caracter incorrecto');
        }
      }
      const reverseCloseChart = arrayClose.reverse();
      for (const index in arrayOpen) {
        if (arrayOpen[index].type !== reverseCloseChart[index].type) {
          setMessage('Error, tienes algun caracter incorrecto');
        }
      }
    }
  }, [openChart, closeChart]);

  const handleChangeText = (t: string) => {
    setText(t);
    setOpenChart([]);
    setCloseChart([]);
    for (let index = 0; index < t.length; index++) {
      if (t[index] === CHART_OPEN_PARENTHESES) {
        setOpenChart(array => [
          ...array,
          {type: PARENTHESES, originalPosition: index},
        ]);
      }
      if (t[index] === CHART_OPEN_BRACKETS) {
        setOpenChart(array => [
          ...array,
          {type: BRACKETS, originalPosition: index},
        ]);
      }
      if (t[index] === CHART_OPEN_CURLY_BRACKETS) {
        setOpenChart(array => [
          ...array,
          {type: CURLY_BRACKETS, originalPosition: index},
        ]);
      }
      if (t[index] === CHART_CLOSE_PARENTHESES) {
        setCloseChart(array => [
          ...array,
          {type: PARENTHESES, originalPosition: index},
        ]);
      }
      if (t[index] === CHART_CLOSE_BRACKETS) {
        setCloseChart(array => [
          ...array,
          {type: BRACKETS, originalPosition: index},
        ]);
      }
      if (t[index] === CHART_CLOSE_CURLY_BRACKETS) {
        setCloseChart(array => [
          ...array,
          {type: CURLY_BRACKETS, originalPosition: index},
        ]);
      }
    }
  };
  return (
    <SafeAreaView>
      <StatusBar />
      <BarText text={text} handleChangeText={handleChangeText} />
      <MessageView message={message} />
    </SafeAreaView>
  );
};

export default App;
