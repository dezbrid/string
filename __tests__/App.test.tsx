/**
 * @format
 */

import 'react-native';
import React from 'react';
import App from '../App';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import {render, fireEvent} from '@testing-library/react-native';

describe('<App />', () => {
  test('renders correctly', () => {
    renderer.create(<App />);
  });
  test('<App /> snapshot', () => {
    const tree = render(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('Should apply the value when changing text (', () => {
    const {getByTestId, findByText} = render(<App />);
    const input = getByTestId('input');
    fireEvent.changeText(input, '(');
    expect(input.props.value).toBe('(');
    expect(findByText('Error, falta caracter de cierre')).toBeTruthy();
  });
  test('Should apply the value when changing text )', () => {
    const {getByTestId, findByText} = render(<App />);
    const input = getByTestId('input');
    fireEvent.changeText(input, ')');
    expect(input.props.value).toBe(')');
    expect(findByText('Error, falta caracter de apertura')).toBeTruthy();
  });
  test('Should apply the value when changing text )(', () => {
    const {getByTestId, findByText} = render(<App />);
    const input = getByTestId('input');
    fireEvent.changeText(input, ')(');
    expect(input.props.value).toBe(')(');
    expect(findByText('Error, tienes algun caracter incorrecto')).toBeTruthy();
  });
  test('Should apply the value when changing text ({])', () => {
    const {getByTestId, findByText} = render(<App />);
    const input = getByTestId('input');
    fireEvent.changeText(input, '({])');
    expect(input.props.value).toBe('({])');
    expect(findByText('Error, tienes algun caracter incorrecto')).toBeTruthy();
  });
});
