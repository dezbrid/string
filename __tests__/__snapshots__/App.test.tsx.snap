// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> <App /> snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
>
  <View
    style={
      Object {
        "margin": 20,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "borderRadius": 20,
            "flexDirection": "row",
            "height": 45,
            "paddingLeft": 10,
            "paddingRight": 15,
          },
          Object {
            "borderColor": "#F1F1F1",
            "borderWidth": 1,
            "shadowColor": "#939393",
            "shadowOffset": Object {
              "height": 1,
              "width": 0,
            },
            "shadowOpacity": 1,
            "shadowRadius": 1,
          },
        ]
      }
    >
      <TextInput
        onChangeText={[Function]}
        placeholder="ingrese el texto aquÃ­"
        placeholderTextColor="#939393"
        style={
          Object {
            "color": "black",
            "flex": 1,
          }
        }
        testID="input"
        value=""
      />
    </View>
  </View>
  <View
    style={
      Object {
        "margin": 20,
      }
    }
  >
    <Text>
      
    </Text>
  </View>
</RCTSafeAreaView>
`;
